processing/src/main/java/io/druid/query/QueryMetrics.java

```diff
@@ -0,0 +1,198 @@
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query;
+
+import com.metamx.emitter.service.ServiceEmitter;
+
+import java.util.Map;
+
+/**
+ * Abstraction wrapping {@link com.metamx.emitter.service.ServiceMetricEvent.Builder} and allowing to control what
+ * metrics are actually emitted, what dimensions do they have, etc.
+ *
+ *
+ * Goals of QueryMetrics
+ * ---------------------
+ *  1. Skipping or partial filtering of particular dimensions or metrics entirely. Implementation could leave the body
+ *  of the corresponding method empty, or implement random filtering like:
+ *  public void cpuTime(long timeNs)
+ *  {
+ *    if (ThreadLocalRandom.current().nextDouble() < 0.1) {
+ *      super.cpuTime(timeNs);
+ *    }
+ *  }
+ *
+ *  2. Ability to add new dimensions and metrics, possibly expensive to compute, or expensive to process (long string
+ *  values, high cardinality, etc.) and not to affect existing Druid installations, by skipping (see 1.) those
+ *  dimensions and metrics entirely in the default QueryMetrics implementations. Users who need those expensive
+ *  dimensions and metrics, could explicitly emit them in their own QueryMetrics.
+ *
+ *  3. Control over the time unit, in which time metrics are emitted. By default (see {@link DefaultQueryMetrics} and
+ *  it's subclasses) it's milliseconds, but if queries are fast, it could be not precise enough.
+ *
+ *  4. Control over the dimension and metric names.
+ *
+ *
+ * Types of methods in this interface
+ * ----------------------------------
+ *  1. Methods, pulling some dimensions from the query object. These methods are used to populate the metric before the
+ *  query is run. These methods accept a single `QueryType query` parameter. {@link #query(Query)} calls all methods
+ *  of this type, hence pulling all available information from the query object as dimensions.
+ *  2. Methods for setting dimensions, which become known in the process of the query execution or after the query is
+ *  completed.
+ *  3. Methods to register metrics to be emitted later in bulk via {@link #emit(ServiceEmitter)}. These methods
+ *  return this QueryMetrics object back for chaining.
+ *
+ *
+ * Implementors expectations
+ * -------------------------
+ * QueryMetrics is expected to be changed often, in every Druid release. Users who create their custom implementations
```

Comment:
I guess this means even a patch release could change QueryMetrics, so this isn't a "stable" extension point. That's fine by me, just want to be clear.